### PR TITLE
Allow specifying the organizationId in the configuration

### LINKED ISSUE #131: Allow for specifying the organization id in the configuration
See https://platform.openai.com/docs/api-reference/organization-optional - allow the organizationId to be specified in the configuration.

### PR DESCRIPTION
**Pull request by edeandrea**
See https://platform.openai.com/docs/api-reference/organization-optional - allow the organizationId to be specified in the configuration.

Fixes #131

### CODE REVIEW

**Comment by edeandrea** [comment]
Just an implementation note. In `OpenAiRestApi` I didn't add an `@ClientHeaderParam` annotation because then I would have had to modify every method signature to add a `@NotBody String organizationId` parameter to it. I thought instead adding a `ResteasyReactiveClientRequestFilter` that conditionally added the header would be a better/cleaner approach.

**Comment by geoand** [comment]
Thanks for this.

I'll have to think about this, I'm not really a fan of the filter for this case.

This might be an opportunity actually to improve the REST Client... 
But like I said, I have to think about it 🙂

**Comment by edeandrea** [comment]
Yeah I'd much prefer to have just the annotation as well, but I thought having to add an extra method parameter to every method, when a lot of the time the value is `null` wasn't the best option either. This way, the filter is only ever added to the chain if the value is even specified, which is better IMO.

**Comment by geoand** [comment]
The thing is that I don't want to have two different approaches for the parameters if it can be avoided.

I'm thinking about some improvements to the `@BeanParam` support of the REST Client (I think we support it, but I could be mistaken) that could apply in this case

**Comment by edeandrea** [comment]
I thought about making the filter more generic as well, something like a `ClientHeaderFilter` where we could configure it with different headers, rather than a specific filter for each set of headers we might want. But for now I kept it simple.

Some of the other edits (like transforming annonymous class instances to lambdas) we done by IntelliJ automatically. I can turn that off and revert those changes if you'd prefer.

**Comment by geoand** [comment]
Yeah, no lambdas please 🙂.

But in any case, I need to look into what I mentioned above before moving this ahead

**Comment by edeandrea** [comment]
> Yeah, no lambdas please

Ok I'll revert that. Can I ask why?

I feel like

```java
return builder::build;
```

is way cleaner than

```java
return new Supplier<>() {
            @Override
            public Object get() {
                return builder.build();
            }
        };
```

and

```java
public void cleanUp(ShutdownContext shutdown) {
     shutdown.addShutdownTask(QuarkusOpenAiClient::clearCache);
}
```

is way cleaner than

```java
    public void cleanUp(ShutdownContext shutdown) {
        shutdown.addShutdownTask(new Runnable() {
            @Override
            public void run() {
                QuarkusOpenAiClient.clearCache();
            }
        });
    }
```

**Comment by geoand** [comment]
> Ok I'll revert that. Can I ask why?

Certainly! 
The reason is that early on during Quarkus' development, we found that when a larger number of lambdas where bootstrapped during application initialization, startup time and RSS took a small hit.

**Comment by geoand** [comment]
Assuming #198 works fine (it works just fine in whatever test I ran), this should be rebased onto it and it will be much simpler

**Comment by edeandrea** [comment]
I'll take a look at that and rebase onto this. Had somewhat of a pet medical emergency today so I might not get to it until tomorrow.

**Comment by geoand** [comment]
Hopefully your pet will be fine!

File: openai/openai-common/runtime/src/main/java/io/quarkiverse/langchain4j/openai/OpenAiRestApi.java [hunk 1]
```diff
@@ -298,6 +300,19 @@ public void aroundWriteTo(WriterInterceptorContext context) throws IOException,
         }
     }
 
+    class OpenAiOrganizationIdRequestFilter implements ResteasyReactiveClientRequestFilter {
```
Comment (geoand) [inline openai/openai-common/runtime/src/main/java/io/quarkiverse/langchain4j/openai/OpenAiRestApi.java:303]: The filter should no longer be necessary now that I added the ApiMetadata class

**Comment by edeandrea** [comment]
Thanks!

So I started rebasing this and saw what you did with the `ApiMetadata` metadata class and `@BeanParam`. Question - this organizationId header is optional, meaning that it shouldn't always be set/sent, and should never be sent in the Azure OpenAI case.

If I add another `@HeaderParam` value in the class, will that value be sent if the value is `null`?

**Comment by geoand** [comment]
A header (or whatever) will not be sent if the value is null

**Comment by edeandrea** [comment]
> A header (or whatever) will not be sent if the value is null

Thats what I needed to know. I'll work on this new integration now.

Just a thought - instead of adding a new method parameter to `public static ApiMetadata of(String apiKey, String apiVersion)` what if I instead created a builder for the `ApiMetaData`? That way in the future whenever we want to add new things we don't have to touch everywhere in the codebase where we use that specific method?

**Comment by geoand** [comment]
Yeah, we can certainly have a builder, but we need to be careful of how the key translates to the two headers

**Comment by edeandrea** [comment]
> Yeah, we can certainly have a builder, but we need to be careful of how the key translates to the two headers

I'll put something together and you can review.

**Comment by geoand** [comment]
👌

**Comment by edeandrea** [comment]
Just pushed changes. Let me know what you think.

File: integration-tests/openai/src/main/java/org/acme/example/openai/QuarkusRestApiResource.java [hunk 2]
```diff
@@ -16,9 +16,7 @@
 */
 package org.acme.example.openai;
 
-import static org.acme.example.openai.MessageUtil.createChatCompletionRequest;
-import static org.acme.example.openai.MessageUtil.createCompletionRequest;
-import static org.acme.example.openai.MessageUtil.createEmbeddingRequest;
+import static org.acme.example.openai.MessageUtil.*;
```
Comment (geoand) [inline integration-tests/openai/src/main/java/org/acme/example/openai/QuarkusRestApiResource.java:19]: I'd prefer we don't use star imports

**Review by geoand — COMMENTED** [review]
I added a minor comment.

Also, it would be nice to have the commits squashed

File: integration-tests/openai/src/main/java/org/acme/example/openai/QuarkusRestApiResource.java (hunk 2 above)
Comment (edeandrea) [reply integration-tests/openai/src/main/java/org/acme/example/openai/QuarkusRestApiResource.java:19]: I ran `mvn impsort:sort` from the project root and thats what it produced. If I commit without running that then the build fails saying I need to run it...

**Review by geoand — APPROVED** [review]
Thanks a lot!

**Comment by geoand** [comment]
I don't *think* I said it's an antipattern.
I definitely did say that for reasons of runtime footprint we avoid lambdas wherever possible

**Comment by edeandrea** [comment]
> I don't _think_ I said it's an antipattern.

No, I don't think you did. I'm saying it is :)

**Comment by edeandrea** [comment]
Not sure why the native java 21 build is failing. Its timing out when creating the binary. I've seen that happen in other apps for java 21. Takes longer to build the image.

**Comment by geoand** [comment]
That's a bug in GraalVM that we need to address.

**Pull request merged**